LLM export pass to swap in custom SDPA

sxu opened the pull request:
Differential Revision: D73444078

jackzhxng commented on extension/llm/export/export_passes.py:
```diff
@@ -95,3 +95,60 @@ def call(self, graph_module: torch.fx.GraphModule):
+    def call_operator(self, op, args, kwargs, meta):
+        from executorch.extension.llm.custom_ops import custom_ops  # noqa
+
+        if op != torch.ops.aten.scaled_dot_product_attention.default:
```
Isn't this op getting decomposed?

sxu replied:
The idea is to run this pass before to_edge, and avoid the decomposed version for perf reasons.

kimishpatel commented on extension/llm/export/export_passes.py:
```diff
@@ -95,3 +95,60 @@ def call(self, graph_module: torch.fx.GraphModule):
+            return super().call_operator(op, args, kwargs, meta)
+
+        q, k, v, *rest = args
+        mask = None
+        dropout = 0.0
+        is_causal = False
+        scale = None
+        if len(rest) > 0:
+            mask = rest[0]
+        if len(rest) > 1:
+            dropout = rest[1]
+        if len(rest) > 2:
+            is_causal = rest[2]
+        if "scale" in kwargs:
+            scale = kwargs["scale"]
```
nit: can be extracted into a small function.

kimishpatel commented on extension/llm/export/export_passes.py:
```diff
@@ -95,3 +95,60 @@ def call(self, graph_module: torch.fx.GraphModule):
+        if op != torch.ops.aten.scaled_dot_product_attention.default:
+            return super().call_operator(op, args, kwargs, meta)
+
+        q, k, v, *rest = args
```
nit: assert that q, k and v are all 4D

kimishpatel commented on extension/llm/export/export_passes.py:
```diff
@@ -95,3 +95,60 @@ def call(self, graph_module: torch.fx.GraphModule):
+        kT = self._transpose(k, meta)
+        vT = self._transpose(v, meta)
+
+        if mask is not None and mask.node.meta["val"].dtype == torch.bool:
```
Put a todo here that custom sdpa once supports boolean mask, this wont be needed. tag me on the todo

kimishpatel commented on extension/llm/export/export_passes.py:
```diff
@@ -95,3 +95,60 @@ def call(self, graph_module: torch.fx.GraphModule):
+                (mask, 0.0, float("-inf")),
+                {},
+                meta,
+            )
```
Also worth checking if the mask is > 2D than add appropriate squeeze ops while making sure first N - 2 dims are all 1

kimishpatel commented on extension/llm/export/export_passes.py:
```diff
@@ -95,3 +95,60 @@ def call(self, graph_module: torch.fx.GraphModule):
+                meta,
+            )
+
+        custom_sdpa = super().call_operator(
```
I would like to add option here that allows us to assume that the mask will be causal and thus we can just set mask =None and is_causal = True, can you do that and add corresponding test?

kimishpatel reviewed (changes requested):
Some nits and special handling for mask and is_causal requested

kimishpatel commented:
cc: @guangy10 @larryliu0820

sxu commented:
@kimishpatel can you take another look?

kimishpatel reviewed (approved):
Thanks for the changes. Looks good. @guangy10 we should look into adopting this as well.

sxu committed:
LLM export pass to swap in custom SDPA (#10355)

Summary: Pull Request resolved: https://github.com/pytorch/executorch/pull/10355

Reviewed By: billmguo, kimishpatel

Differential Revision: D73444078
